### ISSUE TITLE
[FR] Table of contents for documents

### ISSUE THREAD

**Issue by tio-trom** [AppFlowy-IO/AppFlowy#3521]
### Description

Would be very useful to have a table of contents based on the Headings of a document. This way you can structure a document better. This can be added to a right pane.

### Impact

For long documents it is essential to be able to jump to "chapters", that can be represented with "headings" in markdown.

**Comment by LucasXu0** [comment]
@tio-trom Do you mean the outline block?

![image](https://github.com/AppFlowy-IO/AppFlowy/assets/11863087/2c503759-101d-453e-98b6-897148ad644e)

**Comment by tio-trom** [comment]
I mean something like Zettlr has:
![2023-09-25_03-50](https://github.com/AppFlowy-IO/AppFlowy/assets/38043315/c8683ed2-3344-4de3-ab3e-924ad840c3ce)

Where the Headings become "chapters" and when you click one it scrolls to it. If you know what I mean.

**Comment by LucasXu0** [comment]
> Where the Headings become "chapters" and when you click one it scrolls to it. If you know what I mean.

I believe the outline block is precisely what you're looking for, but it's currently displayed on the page instead of in the sidebar.

https://github.com/AppFlowy-IO/AppFlowy/assets/11863087/efc27b34-f542-466f-b7ef-673753997a47

**Comment by tio-trom** [comment]
Ahhh ok...yes that's what I meant, but positioning it like that makes it less efficient if you want to have a quick access you have to always scroll up. Would make more sense in a separate tab or at least a scroll to top icon somewhere.

**Comment by emmggi** [comment]
It makes a lot more sense to me, from a practical standpoint, to have it in a panel. A block would be useful when exporting to a PDF or similar.

**Comment by LucasXu0** [comment]
> Would make more sense in a separate tab or at least a scroll to top icon somewhere.
Sounds reasonable.

I see. @tio-trom @emmggi so do you prefer to display the characters on the panel or the sidebar rather than in the document?

**Comment by tio-trom** [comment]
> > Would make more sense in a separate tab or at least a scroll to top icon somewhere.
> > Sounds reasonable.
> 
> I see. @tio-trom @emmggi so do you prefer to display the characters on the panel or the sidebar rather than in the document?

I think it makes a lot more sense. It is helpful to jump to sections in a document. If you put it at the top you always have to scroll up to then jump to certain sections.

**Comment by emmggi** [comment]
@LucasXu0 I'm not sure about the design. I think having a separate panel for outline that can be toggled on/off makes most sense.

**Comment by annieappflowy** [comment]
@YBYCS , did you mean having the outline floated on a separate panel?

**Comment by dandv** [comment]
Having the table of contents in a panel would be a net advantage over Notion, which only has it at the top of the document.

Consider how much easier it is to navigate among pages in Notion given they are displayed in a tree on the left side. If that tree were at the top and scrolled out, you'd have to scroll up to the top every time you wanted to navigate somewhere.

In the same vein, editors have a "preview" of the file lines in the right pane.

**Comment by Yacine-Benali** [comment]
Yeah this feature is a must for me especially when dealing with big documents. its the only thing preventing me from migrating form obsidian right now

**Comment by xcpky** [comment]
> > Would make more sense in a separate tab or at least a scroll to top icon somewhere.
> > Sounds reasonable.
> 
> I see. [@tio-trom](https://github.com/tio-trom) [@emmggi](https://github.com/emmggi) so do you prefer to display the characters on the panel or the sidebar rather than in the document?

I suggest a toggleable sidebar from the right, similar to the current left side bar.

**Comment by capsudo** [comment]
> > > Would make more sense in a separate tab or at least a scroll to top icon somewhere.
> > > Sounds reasonable.
> > 
> > 
> > I see. [@tio-trom](https://github.com/tio-trom) [@emmggi](https://github.com/emmggi) so do you prefer to display the characters on the panel or the sidebar rather than in the document?
> 
> I suggest a toggleable sidebar from the right, similar to the current left side bar.

That would conflict with [Side Peek](https://github.com/AppFlowy-IO/AppFlowy/issues/8699) view which appears on the right.

IMO would be better to display this ToC as a condensed floating menu that shows headings text only on hover, this is the way Notion implements it:

<img width="296" height="396" alt="Image" src="https://github.com/user-attachments/assets/33f028e2-2d82-4e5e-98e3-776ef0285d11" />

<img width="1355" height="669" alt="Image" src="https://github.com/user-attachments/assets/6a6dbe5a-e8a4-40a4-a84c-7bbc28cc084f" />
